monai/data/png_writer.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2020 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+from skimage import io, transform
+
+
+def write_png(data,
+              file_name,
+              output_shape=None,
+              interp_order=3,
+              mode='constant',
+              cval=0,
+              scale_factor=255):
+    """
+    Write numpy data into png files to disk.  
+    Spatially It supports HW for 2D.(H,W) or (H,W,3) or (H,W,4)
+
+    Args:
+        data (numpy.ndarray): input data to write to file.
+        file_name (string): expected file name that saved on disk.
+        output_shape (None or tuple of ints): output image shape.
+        interp_order (int): the order of the spline interpolation, default is 3.
+            The order has to be in the range 0 - 5.
+            this option is used when `output_shape != None `.
+        mode (`reflect|constant|nearest|mirror|wrap`):
+            The mode parameter determines how the input array is extended beyond its boundaries.
+            this option is used when `output_shape != None`.
+        cval (scalar): Value to fill past edges of input if mode is "constant". Default is 0.0.
+            this option is used when `output_shape != None`.
+    """
+    assert isinstance(data, np.ndarray), 'input data must be numpy array.'
+
+    if scale_factor > 0:
```

Comment:
I think the scale should happen after the resizing (after line 54), as the interpolation might change the original data value range. Also another minor issue is that the method currently couldn't be configured to not scaling `data`, which is a limitation

Reply:
We are only scaling after the resize. So, basically I have to move my conditional statement before multiplying with scale. The first normalization doesn't require that condition. I can do that.

 Also, to your question of no scale, can't we set the scale to 1 to make it use as it is.

Reply:
sorry I didn't make it clear @Bala93, I think there are two minor issues:
- the method's default: if the user just want to save `data` as it is -- by simply `write_png(data, "output.png")`, it seems the current implementation will always scale `data` (which could be a surprise)
- the resize might change `min(data)` and `max(data)`. for example if the input `data` are in the range of [0-255], after resizing (with interpolation order > 1) the pixel values might be outside of [0-255]. so we could do scale after resize, or preserve the value range by doing something like: 
```
min_data = min(data)
max_data = max(data)
data_to_save = clip(resize(data), min_data, max_data)
```

Reply:
Hi @wyli and @Bala93 ,

So in summary, the solution to this question is just to add `preserve_rangebool=True` when calling resize, right?

Reply:
> - the resize might change `min(data)` and `max(data)`. for example if the input `data` are in the range of [0-255], after resizing (with interpolation order >= 1) the pixel values might be outside of [0-255]. 

What you are saying makes sense, but if we do resize there are other problems also: Let's say we have a image with four unique values. When we do a bilinear/bicubic interpolation, the number of unique values now will be more. It won't be just the 4 values. When we do resize apparently we will have this issues. So, now the questions ar: 1) what kind of images are we expecting people to save ? 2) Is output spatial size a necessary meta data ?

Reply:
Hi @Bala93 ,

1. If user doesn't want to generate new unique values, he should choose the right interp mode `reflect|constant|nearest|mirror|wrap`, for example, if choosing `nearest`, no new value generated.
2. This function mainly works for model output, to save 2D segmentation output into file.
3. `spatial_size` is necessary to restore model output as the same size as input image/label. And then put them together to manually evaluate and check bad cases.

Thanks.

Reply:
Ok. I get it. Now, can you please list me the issues which I should resolve ?

Reply:
in summary I think for `write_png` we could:
- add `plugin` and `**plugin_args` as additional optional arguments as they are useful for `imsave` https://scikit-image.org/docs/dev/api/skimage.io.html#imsave
- default to no resizing, no intensity scaling, no datatype conversion. but if they are specified, we do resize first and then rescale and datatype conversion (could make use of `rescale_array` https://github.com/Project-MONAI/MONAI/blob/master/monai/transforms/utils.py#L53)